### ISSUE TITLE
Unmarked Duplicate Reactions From 2 Pdep Networks

### ISSUE THREAD

**Issue by connie** [ReactionMechanismGenerator/RMG-Py#566]
Pdep reaction in chebyshev not properly marked as duplicates when they are coming from different pdep networks..

```
# Data sources
database(
    thermoLibraries = ['primaryThermoLibrary'],
    reactionLibraries = [],
    seedMechanisms = [],
    kineticsDepositories = ['training'], 
    kineticsFamilies = 'default',
    kineticsEstimator = 'rate rules',
)

# Constraints on generated species
generatedSpeciesConstraints(
    maximumCarbonAtoms = 7,
    maximumRadicalElectrons = 1,
)

# List of species
species(
    label='n-heptane',
    structure=SMILES("CCCCCCC"),
)

species(
    label='Ar',
    reactive=False,
    structure=SMILES("[Ar]"),
)

simpleReactor(
    temperature=(1000,'K'),
    pressure=(400,'Pa'),
    initialMoleFractions={
        "n-heptane": 0.02,
        "Ar": 0.98,
    },
    terminationConversion={
        'n-heptane': 0.99,
    },
    terminationTime=(1e6,'s'),
)

simpleReactor(
    temperature=(1500,'K'),
    pressure=(400,'Pa'),
    initialMoleFractions={
        "n-heptane": 0.02,
        "Ar": 0.98,
    },
    terminationConversion={
        'n-heptane': 0.99,
    },
    terminationTime=(1e6,'s'),
)

simpleReactor(
    temperature=(2000,'K'),
    pressure=(400,'Pa'),
    initialMoleFractions={
        "n-heptane": 0.02,
        "Ar": 0.98,
    },
    terminationConversion={
        'n-heptane': 0.99,
    },
    terminationTime=(1e6,'s'),
)

simulator(
    atol=1e-16,
    rtol=1e-8,
)

model(
    toleranceMoveToCore=0.05,
    toleranceInterruptSimulation=0.05
)

pressureDependence(
    method='modified strong collision',
    maximumGrainSize=(0.5,'kcal/mol'),
    minimumNumberOfGrains=250,
    temperatures=(300,3000,'K',8),
    pressures=(0.001,100,'bar',5),
    interpolation=('Chebyshev', 6, 4),
)


```

rmgpy.chemkin.ChemkinError: Encountered unmarked duplicate reaction C4H5(156) + H(7) <=> CH3(10) + C3H3(100).

It appears that they are bimolecular, which may be why they don't process correctly in `processNewReactions()`

**Comment by connie** [comment]
More investigation: 

The reaction starts out correctly labeled as a duplicated in chem0037.inp:

`CH3(10)+C3H3(100)(+M)=>C4H5(156)+H(7)(+M)`

but from chem0038 on, it gets reversed and now is no longer unidirectional but bidirectional:

`C4H5(156)+H(7)(+M)=CH3(10)+C3H3(100)(+M)`

Also loses the duplicate flag..

**Comment by connie** [comment]
I suspect the following lines are the culprit:

https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/rmgpy/rmg/model.py#L816

and the function https://github.com/ReactionMechanismGenerator/RMG-Py/blob/master/rmgpy/chemkin.py#L1666

We presuppose that all the core reactions are already checked, but somehow because they are in the pdep network, they have lost their duplicate tags due to shifting from unidirectional to bidirectional....

**Comment by nickvandewiele** [comment]
I think @alongd can share similar experiences on duplicate p-dep reactions.

**Comment by connie** [comment]
I asked shamel about this, and it seems these two reactions should be in the same pdep network.  In other words, RMG-Py is not merging the two networks properly, whereas he thinks that RMG-Java does merge them when all the products are found in both.  So probably solving the mark duplicates is not a completely sufficient solution for fixing the chemistry.

**Comment by connie** [comment]
I have just confirmed that pressure dependence in RMG-Java does not produce any duplicates for Net Reactions from pdep networks.  Appears that Py is definitely not merging the networks properly.

**Comment by alongd** [comment]
I got similar bi-molecular dups working with PDepArrhenius (not only Chebyshev, about 10 dups per run).
Can the one-way reactions be calculated as reversible reactions first thing as they are added to the pdep network?

**Comment by connie** [comment]
Just to confirm, were the PDepArrhenius reactions generated by RMG rather than from a seed mechanism or reaction library?

I don't know the answer to your question...

**Comment by alongd** [comment]
Yes, the Pdep reactions were generated by RMG.
The original reaction could come either from a library (e.g. Glarborg C3) of family (e.g. Disproportionation), and is not pressure dependent. The second reaction comes from the PdepNetwork.

**Comment by rwest** [comment]
This is all quite hand-wavey because I've not thought in depth about this at all recently, just trying to share some hunches and vague recollections in case they help...  I think there may be something funny about "leak" reactions from PDep networks. Like the master equation is solved for each entrance channel, which would presumably only include core species, and gives the net rates to any exit channels that are also in the core, but the "leak" rates to exit channels that are not in the model core or have not been "explored" for that PDep network. Leak reactions to edge species perhaps should be irreversible, but I think anything ending up in the core should probably be fully explored and reacting reversibly.  There's definitely something slightly confusing about terminology "core, edge, explored, leak, isomer" etc. which all mean specific and different things. If I recall correctly, I think even @jwallen was confused about the Matheu way of exploring and merging PDep networks for a while, and for some time RMG-Py did it quite differently from RMG-Java.

**Comment by connie** [comment]
I do think what we are currently seeing is probably incorrect:

We have in PDepNetwork 1 with reactants A, B, C, D and isomer I.  with reactions A+B = I and C+D = I, bringing in the net reaction A+B = C+D

Then there is a PDepNetwork 2 with reactants A, B, C, D and isomer X with reactions A+B = X and C+D = X, bringing in the same net reaction A+B = C+D.  

It would make sense for both isomers I and X to appear in the same PDep network.

**Comment by jwallen** [comment]
I think @connie's explanation above is plausible, that RMG could find PDepNetworks for A + B <=> X* <=> C + D and A + B <=> Y* <=> C + D without finding X* <=> ... <=> Y*, meaning the two networks would never be combined and you'd get duplicate net reactions A + B <=> C + D. If this is the explanation, I see two possible solutions:

1. Detect when this occurs, then merge the networks by exploring isomers as needed. This would be the most robust, but would require some careful tweaking of the algorithm to efficiently identify duplicate PDepReactions and explore the corresponding PDepNetworks. If X* and Y* are multiple transition states apart, this could result in a very large PDepNetwork.

2. Accept the duplicate reactions (marking them as such). This would work as long as there is code to handle the possibility that the networks get merged in a later iteration of the mechanism generation, which would cause a reaction to disappear from the edge or core of the model.

I'd lean towards option 2, but before implementing either I'd suggest rigging up a toy example to see if the sum of the k(T,P) for the A + B <=> C + D net reactions from the separate PDepNetworks is at all close to the k(T,P) for the A + B <=> C + D net reaction from the merged PDepNetwork. I think they should be; if they aren't we'd want to think about what that means in terms of the underlying master equation model.

**Comment by mliu49** [comment]
I think this is mostly addressed by #1296.
